### ISSUE TITLE
Cleaning up a request-response cycle regardless of errors

### ISSUE THREAD

**Issue by gitblit** [decebals/pippo#81]
Consider this scenario where we are getting a Db instance to be shared throughout a request-response cycle:

```java
// Preparation route for 
ALL("/.*", (request, response, chain) -> {
    response.bind("db", getDb());
}); 

POST("/{id}", (request, response, chain) -> {
    Db db = (Db) response.getLocals().get("db");
    // this throws an runtime exception somewhere deep inside
    db.insert(something);
});

// Cleanup route
ALL("/.*", (request, response, chain) -> {
    Db db = (Db) response.getLocals().remove("db");
    db.release();
}); 
```

If an exception is thrown in `POST` then the cleanup route is never executed and we are now leaking Db connections.  Instead of trapping Db exceptions in every handler, it would be nice to be able to specify cleanup routes/routines that are guaranteed to execute at the end of the Request-Response cycle **despite** exceptions within the chain.  These mechanisms should be fail-safe themselves and should not have a `chain` parameter.

```java
void cleanup(Request request, Response response);
```

The `RouteHandlerChain` or `PippoFilter` should execute these cleanup routines within independent `try {} catch {}` blocks.

```java
for (CleanupHandler handler : cleanupHandlers) {
    try {
        handler.cleanup(request, response);
    } catch (Exception e) {
        log.error("Whoops", e);
    }
}
```

Perhaps these could be registered as:

```java
// Cleanup route
CLEANUP("/.*", (request, response) -> {
    Db db = (Db) response.getLocals().remove("db");
    db.release();
}); 
``````

**Comment by decebals** [comment]
Personal, I found two options:
- use a `RouteHandler` (filter - no commit) and a custom `ErrorHandler`
```java
setErrorHandler(new DefaultErrorHandler() {

    @Override
    public void handle(Exception exception, Request request, Response response) {
        >> here close database connection
        super(exception, request, response);
    }

}

>> here add routes

// Cleanup route
ALL("/.*", (request, response, chain) -> {
    Db db = (Db) response.getLocals().remove("db");
    db.release();
}); 
```
- add a marker __Finally__ for `RouteHandler` (`Finally` annotation or `FinallyRouteHandler` that extends `RouteHandler`) and in PippoFilter when we catch an exception in a RouteHandler then iterates through the chain, find FinallyRouteHandler and call them.

By the way. I think that it's better to change this code from PippoFilter
```java
} catch (Exception e) {
    log.error(e.getMessage(), e);
    if (!response.isCommitted()) {
        errorHandler.handle(e, request, response);
    } else {
        log.debug("The response has already been committed. Cannot use the exception handler.");
    }
}
```

with a simple
```java
} catch (Exception e) {
    log.error(e.getMessage(), e);
    errorHandler.handle(e, request, response);
}
```
and move the if/else block in (Default)ErrorHandler. In this mode we are more flexible.

What you think?

**Comment by gitblit** [comment]
The first suggestion is how it could be done right now without API changes.  I don't care for this approach as I don't find it to be terribly elegant.

The second approach is more to my liking.  Annotations are nice.  That would work with a controller-based design but it would not work with Lambdas.  Hard classes are nice.  Would work with Lambdas but not controllers.  Do we need both?

As for the error handler change, sure as long as `DefaultErrorHandler` performs that check.

**Comment by balamaci** [comment]
Hi guys, I don't know if a CLEANUP or FINALLY annotation is very clear for a new user it exists and it needs to use that. I remember @gitblit mentioning something that stuck: "the rule of least astonishment" and I guess it also applies to the custom number of things a new user needs to learn to use a new framework. I think we need to consider how often we expect the users to use this cleanup pattern before doing some custom things. I'm thinking you are referring to something alike "OpenSessionInView" or a MDC value kept in ThreadLocal. 
Maybe something like a request listener(onbegin and onfinish) could be considered to support this?

**Comment by gitblit** [comment]
A listener design would be another option.  I could go either way for my needs.

There are two nice things about the `RouteHandler` approach:

1. we are not adding a new mechanism with registration/removal methods
2. the appropriate cleanup/finally handler would be matched with `uriPattern` by the `Router`.

A listener mechanism would either need to also match by `uriPattern` or would be a global mechanism for all requests.  Having a global `onFinished` request listener could have other creative uses.

Like I said, either would solve the scenario I've outlined.  One is more targeted than the other.

I guess the real question should be: is cleanup/finally a filter or is it a listener?

**Comment by decebals** [comment]
It's important for adoption to have a steep learning curve. For this reason I wish to not introduce new concepts if not necessarily need.
> I guess the real question should be: is cleanup/finally a filter or is it a listener?

I prefer a filter.
I propose an idea introduced by James (gitblit) in https://github.com/decebals/pippo/issues/66.
The idea consists in:
- returns a Route in `Application.ALL` (in don't know if it's ok to make this modification for other methods - GET for example)
- add a `cleanup/finally` method to Route

```java
ALL("/.*", (request, response, chain) -> {
    >> here close database connection
}).cleanupOrFinally();
```

Make sense?

**Comment by gitblit** [comment]
Yup.  That'll work when coupled with appropriate logic in PippoFilter.

**Comment by balamaci** [comment]
Very nice. Maybe rephrase the method name as to suggest more of a flag setting, not an actual action.
I guess passing a lambda instead to cleanup( ()-> { .. }) is not a better option.

**Comment by decebals** [comment]
@balamaci I think that can be an idea (a finally block for a concrete route) but the implementation requires java 8 and for now I say to stay with java 7 for a while.

I created a new branch  [finally](https://github.com/decebals/pippo/tree/finally) to create a RFC for this feature.
I added new routes in `BasicApplication` to make some tests:
```java
// throw a programatically exception
GET("/exception", new RouteHandler() {

    @Override
    public void handle(Request request, Response response, RouteHandlerChain chain) {
        throw new RuntimeException("My programatically error");
    }

});

// use a finally filter (invoked even when exceptions were raised in previous routes)
// test with route "/" and "/exception"
ALL("/.*", new RouteHandler() {

    @Override
    public void handle(Request request, Response response, RouteHandlerChain chain) {
        System.out.println(">>> Cleanup here");
    }

}).runAsFinally();
```

In the internet browser I made a request to `http://localhost:8338/exception` and I noticed that the cleanup filter is executed (for the moment I implemented nothing in `PippoFilter` related to `runAsFinally` flag). 

I will inform you about evolution of this branch.

**Comment by decebals** [comment]
> In the internet browser I made a request to http://localhost:8338/exception and I noticed that the cleanup filter is executed (for the moment I implemented nothing in PippoFilter related to runAsFinally flag).

After some tests I can infirm this affirmation. 

I made some tests and I found some issues in actual pippo-core:
- if I call "/not-found" from browser then "cleanup" filter is called because the URI pattern is `ALL("/.*")` (here I expect route not found)
- I see a request "/favicon.ico" in pippo from browser (I resolved with 5c1c43d7318ff8f92c9848831de0868f39125d2f) that activate the "cleanup" filter and I don't want this behavior

Questions:
- how we can solve first issue described above ("/not-found")?
- do I want to call the "cleanup" filter if chain.next() was not called in previous route?
- is it possible to write (and how) some unit tests for this feature (because I think that)?

**Comment by gitblit** [comment]
In PippoFilter after the ignore paths check we should...

```java
response.status(0);
```

Then after the check for `!response.isCommitted()` we should...

```java
if (response.getStatus() == 0) {
    response.notFound();
    log.info("Handlers in chain did not set a status code for {} '{}'",
requestMethod, requestUri);
}
```

I implemented that last night but didn't propose it yet.  Works nicely.

Because of filters, we have to default to an uninitialized status code.  A
handler somewhere in the chain must set a status code, otherwise it's a 404.

-J

**Comment by gitblit** [comment]
I pushed a change to your `finally` branch to handle the `/not-found` scenario.  We couldn't use 0 on Jetty, so I used Integer.MAX_VALUE instead.  Some handler in the chain must set a status code, otherwise we trigger a 404.

**Comment by gitblit** [comment]
I was not able to repeat your `runAsFinally()` observation - i.e. I disagree that that cleanup occurred after an exception was thrown.  Perhaps you were testing with the `/not-found` path which should have been a 404 but was not and also did not throw an exception.

I completed your implementation of `runAsFinally()` by integrating it into the  into the `RouteHandlerChain` and ensuring it's called from the finally block of `PippoFilter`.  It is now working as envisioned.

Side note, I find the IntelliJ Grep Console plugin to be very helpful in visually filtering out signal from noise.  It automatically colors debug, info, warn, etc differently.

**Comment by decebals** [comment]
> I was not able to repeat your runAsFinally() observation - i.e. I disagree that that cleanup occurred after an exception was thrown. Perhaps you were testing with the /not-found path which should have been a 404 but was not and also did not throw an exception.

It was my mistake. 

I like IntelliJ Grep Console plugin. 

Can we find a better name for `runAsFinally` or it's ok? Now in `RouteHandlerChain` we have `cleanup` and maybe it's good to use the same term to mark a cleanup route.

**Comment by gitblit** [comment]
I like both finally and cleanup.  Unfortunately "finally", by itself, is a
reserved keyword.  I'm happy with whichever you prefer.

**Comment by decebals** [comment]
Just an observation. Now I see that I can call `chain.cleanup()` (accidentally or not) in a route handler. Is it ok?

**Comment by gitblit** [comment]
Resolved by #84.
